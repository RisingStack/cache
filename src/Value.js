/* @flow */

export default class Value<V> {
  value: V;
  options: TTLOptions;
  createdAt: Date;
  expireAt: Date;
  staleAt: Date;

  constructor(value: V, options: TTLOptions = {}) {
    this.value = value
    this.refresh(options)
  }

  refresh(options: TTLOptions = {}) {
    this.options = options
    const { expire, stale, createdAt = Date.now() } = options
    this.createdAt = new Date(createdAt)

    if (Number.isInteger(expire) && expire !== 0) {
      this.expireAt = new Date(createdAt + expire)
    }

    if (Number.isInteger(stale) && stale !== 0) {
      this.staleAt = new Date(createdAt + stale)
    }
  }

  isExpired(): boolean {
    if (!this.expireAt) {
      return true
    }

    return this.expireAt.valueOf() < Date.now()
  }

  isStale(): boolean {
    if (!this.staleAt) {
      return true
    }

    return this.staleAt.valueOf() < Date.now()
  }

  toJSON(): string {
    return JSON.stringify({
      value: this.value,
      options: Object.assign({
        createdAt: this.createdAt.valueOf(),
        stale: undefined,
        expire: undefined
      }, this.options)
    })
  }

  static fromJSON(json: string): ?Value<V> {
    try {
      const { value, options } = JSON.parse(json)
      if (!options.createdAt) {
        return null
      }

      return new Value(value, options)
    } catch (err) {
      return null
    }
  }
}
