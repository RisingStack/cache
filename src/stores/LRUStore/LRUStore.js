/* @flow */

import LRU from 'lru-cache'
import Value from '../../Value'

type LRUCache<V> = {
  itemCount: number;
  set(key: string, value: V, maxAge: ?number): null;
  get(key: string): V;
  del(key: string): null;
  reset(): null;
}

export default class MemoryStore<V> {
  store: LRUCache<Value<V>>;
  stats: StoreStats;

  constructor(options: number | { max: number }) {
    this.store = LRU(options)
    this.stats = {
      getCount: 0,
      hitCount: 0
    }
  }

  get(key: string): Promise<?Value<V>> {
    const value = this.store.get(key)

    this.stats.getCount += 1
    if (value && (!value.isExpired() || !value.isStale())) {
      this.stats.hitCount += 1
    }

    return Promise.resolve(value)
  }

  set(key: string, value: V, options: TTLOptions = {}): Promise<Value<V>> {
    const val = new Value(value, options)
    this.store.set(key, val, options.expire)
    return Promise.resolve(val)
  }

  delete(key: string): Promise<*> {
    this.store.del(key)
    return Promise.resolve()
  }

  clear() {
    this.store.reset()
  }

  async size(): Promise<number> {
    return Promise.resolve(this.store.itemCount)
  }

  getStats(): StoreStats {
    return this.stats
  }

  resetStats(): StoreStats {
    this.stats = {
      getCount: 0,
      hitCount: 0
    }

    return this.getStats()
  }
}
