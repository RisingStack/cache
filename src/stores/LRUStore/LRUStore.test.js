import LRUStore from './LRUStore'
import Value from '../../Value'

describe('LRUStore', () => {
  let store

  beforeEach(() => {
    store = new LRUStore()
  })

  describe('.get()', () => {
    it('should return a Promise', async () => {
      let value = await store.get('notexisting')
      expect(value).toBeUndefined()

      const item = { val: 1 }
      store.set('existing', item)

      value = await store.get('existing')
      expect(value).toBeInstanceOf(Value)
      expect(value).toMatchObject({ value: item })
      expect(store.store.dump().map(({ k, v }) => ({ k, v }))).toMatchSnapshot()
    })
  })

  describe('.set()', () => {
    it('should set a value with ttl options', async () => {
      const item = { val: 1 }
      const options = { stale: 5, expire: 10 }
      store.set('key', item, options)

      const value = await store.get('key')
      expect(value).toMatchObject({
        options,
        value: item
      })
      expect(store.store.dump().map(({ k, v }) => ({ k, v }))).toMatchSnapshot()
    })
  })

  describe('.delete()', () => {
    it('should delete an item', async () => {
      const item = { val: 1 }
      store.set('key', item)

      let value = await store.get('key')
      expect(value).toBeDefined()

      store.delete('key')

      value = await store.get('key')
      expect(value).toBeUndefined()
      expect(store.store.dump().map(({ k, v }) => ({ k, v }))).toMatchSnapshot()
    })
  })

  describe('.clear()', () => {
    it('should clear the store', async () => {
      const item = { val: 1 }
      store.set('key', item)

      let value = await store.get('key')
      expect(value).toBeDefined()

      store.clear()

      value = await store.get('key')
      expect(value).toBeUndefined()
      expect(store.store.dump().map(({ k, v }) => ({ k, v }))).toMatchSnapshot()
    })
  })

  describe('.size()', () => {
    it('should return the size of the store', async () => {
      let size = await store.size()
      expect(size).toEqual(0)

      store.set('key', { val: 1 })
      store.set('key2', { val: 1 })

      size = await store.size()
      expect(size).toEqual(2)

      store.delete('key')

      size = await store.size()
      expect(size).toEqual(1)
    })
  })

  describe('.getStats()', () => {
    it('should return hit stats', async () => {
      store.set('existing', { val: 1 })

      await store.get('notexisting')
      await store.get('notexisting')
      await store.get('existing')
      await store.get('notexisting')
      await store.get('existing')

      const stats = store.getStats()
      expect(stats).toEqual({
        getCount: 5,
        hitCount: 2
      })
    })
  })

  describe('.resetStats()', () => {
    it('should reset the hit stats', async () => {
      store.set('existing', { val: 1 })

      await store.get('notexisting')
      await store.get('notexisting')
      await store.get('existing')
      await store.get('notexisting')
      await store.get('existing')

      store.resetStats()

      let stats = store.getStats()
      expect(stats).toEqual({
        getCount: 0,
        hitCount: 0
      })

      await store.get('notexisting')
      await store.get('existing')

      stats = store.getStats()
      expect(stats).toEqual({
        getCount: 2,
        hitCount: 1
      })
    })
  })
})
