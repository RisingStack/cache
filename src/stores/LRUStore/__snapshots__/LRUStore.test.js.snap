// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LRUStore .clear() should clear the store 1`] = `Array []`;

exports[`LRUStore .delete() should delete an item 1`] = `Array []`;

exports[`LRUStore .get() should return a Promise 1`] = `
Array [
  Object {
    "k": "existing",
    "v": "{\\"value\\":{\\"val\\":1},\\"options\\":{\\"createdAt\\":946684800000}}",
  },
]
`;

exports[`LRUStore .set() should set a value with ttl options 1`] = `
Array [
  Object {
    "k": "key",
    "v": "{\\"value\\":{\\"val\\":1},\\"options\\":{\\"createdAt\\":946684800000,\\"stale\\":5,\\"expire\\":10}}",
  },
]
`;
